com.unity.render-pipelines.core/Editor/Lighting/ProbeVolume/ProbeGIBaking.Dilate.cs

```diff
@@ -61,35 +62,93 @@ internal void FromSphericalHarmonicsL2(SphericalHarmonicsL2 sh)
+                    L2_2 = new Vector3(shData[idx0 + 3].z, shData[idx0 + 4].z, shData[idx0 + 5].z);
+                    L2_3 = new Vector3(shData[idx0 + 3].w, shData[idx0 + 4].w, shData[idx0 + 5].x);
+                    L2_4 = shData[idx0 + 6];
+                }
+                else
+                {
+                    L2_0 = L2_1 = L2_2 = L2_3 = L2_4 = Vector3.zero;
+                }
+            }
+
+            internal void ToSphericalHarmonicsShaderConstants(ProbeVolumeSHBands srcBands, NativeArray<Vector4> shData, int probeIdx)
+            {
+                var stride = srcBands == ProbeVolumeSHBands.SphericalHarmonicsL2 ? 7 : 3;
+                var idx0 = probeIdx * stride;
+
+                shData[idx0 + 0] = new Vector4(L0.x, L0.y, L0.z, L1_0.x);
+                shData[idx0 + 1] = new Vector4(L1_0.y, L1_1.y, L1_2.z, L1_0.y);
+                shData[idx0 + 2] = new Vector4(L1_0.z, L1_1.y, L1_2.z, L1_0.z);
+
+                if (srcBands == ProbeVolumeSHBands.SphericalHarmonicsL2)
+                {
+                    shData[idx0 + 3] = new Vector4(L2_0.x, L2_1.x, L2_2.x, L2_3.x);
+                    shData[idx0 + 4] = new Vector4(L2_0.y, L2_1.y, L2_2.y, L2_3.y);
+                    shData[idx0 + 5] = new Vector4(L2_0.z, L2_1.z, L2_2.z, L2_3.z);
+                    shData[idx0 + 6] = new Vector4(L2_4.x, L2_4.y, L2_4.z, 1f);
+                }
+            }
         }
 
         struct DataForDilation
         {
-            public ComputeBuffer validityBuffer;
-            public ComputeBuffer positionBuffer;
-            public ComputeBuffer outputProbes;
+            public ComputeBuffer validityBuffer { get; }
+            public ComputeBuffer positionBuffer { get; }
+            public ComputeBuffer outputProbes { get; }
+
+            DilatedProbe[] dilatedProbes;
+
+            ProbeReferenceVolume.Cell cell;
 
             public DataForDilation(ProbeReferenceVolume.Cell cell)
             {
+                this.cell = cell;
+
                 int probeCount = cell.probePositions.Length;
 
                 validityBuffer = new ComputeBuffer(probeCount, sizeof(float));
                 positionBuffer = new ComputeBuffer(probeCount, System.Runtime.InteropServices.Marshal.SizeOf<Vector3>());
                 outputProbes = new ComputeBuffer(probeCount, System.Runtime.InteropServices.Marshal.SizeOf<DilatedProbe>());
 
-
                 // Init with pre-dilated SH so we don't need to re-fill from sampled data from texture (that might be less precise).
-                DilatedProbe[] dilatedProbes = new DilatedProbe[probeCount];
+                dilatedProbes = new DilatedProbe[probeCount];
                 for (int i = 0; i < probeCount; ++i)
                 {
-                    dilatedProbes[i].FromSphericalHarmonicsL2(cell.sh[i]);
+                    dilatedProbes[i].FromSphericalHarmonicsShaderConstants(cell.shBands, cell.shData, i);
```

Comment:
The problem I have with this not dilating the whole 3 bands is that this now makes switching between L1 and L2 at runtime not possible. 

Let's assume the data is baked with L1 set as option, then it will be stored as L1 with garbage L2. 
Switching then to L2 will require a rebake as otherwise garbage will be read. 

I see where you are coming from, but I can also this potentially rising issues and annoyances for users.

Reply:
Before I go on and look more at the PR we should discuss this as I think is pivotal for the rest.

Reply:
As discussed, we currently always store all three bands when baking (i.e. ignore current lighting setting for bake). A new toggle should be introduced for controlling how many bands to store / how many bands to bundle with builds. Loading mixed cells already works.